### PR TITLE
[MIG]Migrated Business requirement module from 8.0 to 10.0

### PR DESCRIPTION
**Pull request by YogeshMahera-SerpentCS**
#48 

- Migrated Business Requirement from v8 to v10

### CODE REVIEW

**Comment by YogeshMahera-SerpentCS** [comment]
@elicoidal @victormartinelicocorp 
Please review it.

**Comment by martintorresvm** [comment]
@elicoidal Should we keep this open to include demo data from #212 ?

**Comment by elicoidal** [comment]
@victormartinelicocorp We will merge the PR in v8 in short so that we can push forward

**Review by elicoidal — COMMENTED** [review]
LGTM.
Demo data to be added once merged in v8.
README to be updated.
Functional test to be done

**Comment by elicoidal** [comment]
Coverage is low: let's try to bring it up to 90% at least

**Comment by elicoidal** [comment]
Looks like a first good start. Some small issues to be corrected:
- [x] when creating a BR, the focus is in the use case field. should be in name
- [x] only graph view available: pivot should be added as well
- [x] there is no report available: at the basic one should be available (Business Requirement Report)
- [x] Source field is too wide: could it be set as one size high (now it seems 2)

**Comment by elicoidal** [comment]
It is strange: normally changing the stages is done by clicking on the stage itself (latest developments) but it seems that the buttons are back in this version.
@YogeshMahera-SerpentCS can you check it?

**Comment by elicoidal** [comment]
A new request: I want that some of the modules appear as Application in the list:
- [x] create a "Business Requirement" Application
- [x] set the following modules as Apps: Business Requirement, Business Requirement Deliverable and Business Requirement Project

**Comment by elicoidal** [comment]
- [x] Source Field should be set as read-only as of Confirmed state (to be backported to v8 too)

**Comment by YogeshMahera-SerpentCS** [comment]
`only graph view available: pivot should be added as well`

- Which field should be added in pivot view or same graph view fields?

`there is no report available: at the basic one should be available (Business Requirement Report)`

- In v8 there is no report like Business Requirement Report, so if is it require then what should be format?.

@elicoidal

**Comment by elicoidal** [comment]
@YogeshMahera-SerpentCS 
>Which field should be added in pivot view or same graph view fields?

When I install the modules in v8, I have a pivot view for the "Business Requirement". Maybe it is defined in another module. Please check current Odoo 8 version
> In v8 there is no report like Business Requirement Report, so if is it require then what should be format?.

Yes there is: please check current Odoo 8 version

**Comment by YogeshMahera-SerpentCS** [comment]
> When I install the modules in v8, I have a pivot view for the "Business Requirement". Maybe it is defined in another module. Please check current Odoo 8 version

- Yes, I see in v10 there is pivot view is itself record view. In v8 it's included in graph view type as `pivot`.

> Yes there is: please check current Odoo 8 version
- Yes, it's available in `business_requirement_deliverable_report`.so shall I copy from `business_requirement_deliverable_report` to `business_requirement` module.?

@elicoidal

**Comment by elicoidal** [comment]
Yes: the one for BR should be moved here and the one for deliverable in br_deliverable and the one for cost in brd_cost module
We should have done it in v8... but it is OK

**Comment by YogeshMahera-SerpentCS** [comment]
@elicoidal 

> It is strange: normally changing the stages is done by clicking on the stage itself (latest developments) but it seems that the buttons are back in this version.
@YogeshMahera-SerpentCS can you check it?

Yes, I have checked its code in `business_requirement_deliverable_project`.
so shall I do stage as itself clicking in BR main module?

**Comment by elicoidal** [comment]
@YogeshMahera-SerpentCS Please fix v8 and we will forward port to v10 for the sake of consistency

**Comment by elicoidal** [comment]
@YogeshMahera-SerpentCS for pivot view: please do not set any default dimension.

**Comment by YogeshMahera-SerpentCS** [comment]
@elicoidal @victormartinelicocorp Could you please review.

**Review by elicoidal — APPROVED** [review]
Functional tests OK.
I will work on the README refreshing in short

**Comment by elicoidal** [comment]
@victormartinelicocorp @seb-elico please have a look

File: business_requirement/models/business.py [hunk 1]
```diff
@@ -199,18 +196,23 @@ def _get_default_company(self):
...
         string='To be Reviewed'
     )
-    kanban_state = fields.Selection([('normal', 'In Progress'),
-                                     ('on_hold', 'On Hold'),
-                                     ('done', 'Ready for next stage')],
-                                    'Kanban State',
-                                    track_visibility='onchange',
-                                    required=False,
-                                    copy=False, default='normal')
+    kanban_state = fields.Selection(
```
Comment (martintorresvm) [inline business_requirement/models/business.py:199]: @elicoidal @seb-elico Too late for ask use https://github.com/OCA/server-tools/tree/10.0/base_kanban_stage
to add this feature?

Forgot about this comment

Comment (elicoidal) [reply business_requirement/models/business.py:199]: make your case :wink:

File: business_requirement/tests/test_br.py [hunk 2]
```diff
@@ -101,9 +104,29 @@ def test_br_name_search(self):
             'parent_id': False,
         }
         self.br.create(br_vals)
+
         brs = self.br.name_search(name='test')
         self.assertEqual(bool(brs), True)
 
+    def test_create_sequance(self):
```
Comment (martintorresvm) [inline business_requirement/tests/test_br.py]: r/test_create_sequance
s/test_create_sequence
@YogeshMahera-SerpentCS Where is checked the sequence created? I didn't see any assert

File: business_requirement/report/report.xml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+   <report
```
Comment (seb-elico) [inline business_requirement/report/report.xml]: Wrong indent of attributes according to OCA guidelines

File: business_requirement/security/security.xml [hunk 4]
```diff
@@ -1,34 +1,32 @@
...
-            <field name="name">Business Requirement</field>
-            <field name="model_id" ref="model_business_requirement"/>
-            <field name="global" eval="True"/>
-            <field name="domain_force">['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
-        </record>
+    <record id="rule_business_requirement" model="ir.rule">
+        <field name="name">Business Requirement</field>
+        <field name="model_id" ref="model_business_requirement"/>
+        <field name="global" eval="True"/>
+        <field name="domain_force">['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
```
Comment (seb-elico) [inline business_requirement/security/security.xml]: Add spaces after each `,` to improve readability

File: business_requirement/views/business_view.xml [hunk 5]
```diff
@@ -1,329 +1,247 @@
...
+                <header>
+                    <field name="state" widget="statusbar" clickable="True"
+                           statusbar_visible="draft,confirmed,approved,stakeholder_approval,in_progress,done,cancel,drop"/>
+                </header>
+                <sheet>
+                    <div class="oe_left">
+                        <field name="kanban_state" class="oe_inline" widget="kanban_state_selection"/>
+                    </div>
+                    <div class="oe_right oe_button_box" name="buttons" groups="base.group_user">
+                         <button class="oe_inline oe_stat_button" type="action" name="%(act_view_sub_br_all)d" icon="fa-tasks">
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Could you please reduce the length of this line?

File: business_requirement/views/business_view.xml [hunk 6]
```diff
@@ -1,329 +1,247 @@
...
+            </tree>
+        </field>
+    </record>
+
+    <record id="view_business_requirement_search" model="ir.ui.view">
+        <field name="name">business.requirement.search</field>
+        <field name="model">business.requirement</field>
+        <field name="arch" type="xml">
+            <search string="Business Requirement">
+                <field name="name" string="Name" filter_domain="['|',('name','ilike',self),('description','ilike',self)]"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Could you please improve the format of this line?
Could you please reduce the length of this line?

File: business_requirement/views/business_view.xml [hunk 7]
```diff
@@ -1,329 +1,247 @@
...
+                <field name="project_id" string="Project"/>
+                <field name="partner_id" string="Customer"/>
+                <field name="responsible_id" string="Responsible"/>
+                <field name="origin" string="Source"/>
+                <filter string="To be Reviewed" name="to_be_reviewed" domain="[('to_be_reviewed','!=', False)]"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Add space after `,`

File: business_requirement/views/business_view.xml [hunk 8]
```diff
@@ -1,329 +1,247 @@
...
+                <group expand="0" string="Group By">
+                    <filter string="Project" context="{'group_by':'project_id'}"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Add space after `:` (same for next lines)

File: business_requirement/views/business_view.xml [hunk 9]
```diff
@@ -1,329 +1,247 @@
...
+    <record model="ir.actions.act_window" id="action_business_requirement_category_tree">
+        <field name="name">Categories</field>
+        <field name="res_model">business.requirement.category</field>
+        <field name="view_mode">tree,form,graph</field>
+        <field name="view_type">form</field>
+        <field name="context">{}</field>
+    </record>
+
+    <!-- Top menu item -->
+    <menuitem id="menu_business_root"  name="Business Requirement"  groups="business_requirement.group_business_requirement_user" sequence="30"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Could you please reduce the length of these lines?

File: business_requirement/views/business_view.xml [hunk 10]
```diff
@@ -1,329 +1,247 @@
...
+
+    <menuitem id="menu_business_requirement" name="Business Requirement" groups="business_requirement.group_business_requirement_user" parent="menu_business_root" sequence="8"/>
+
+    <menuitem id="menu_business_configuration" name="Configuration" groups="business_requirement.group_business_requirement_user" parent="menu_business_root" sequence="8"/>
+
+    <menuitem id="menu_business_requirement1" groups="business_requirement.group_business_requirement_user" action="action_business_requirement_tree"
+        parent="menu_business_requirement" sequence="10"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Could you please fix the indentation of this line and the next ones?

**Comment by elicoidal** [comment]
Actually I do not see the kanban states @YogeshMahera-SerpentCS

File: business_requirement/views/report_business_requirement.xml [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        </div>
+        <t t-raw="0"/>
+    </template>
+
+    <template id="layout_br_base">
+        <h3><span t-field="o.name"/>-
+            <span t-field="o.description"/>
+        </h3>
+        <table class="table table-bordered"
+               style="table-collapse:collapse; table-layout:fixed">
```
Comment (seb-elico) [inline business_requirement/views/report_business_requirement.xml]: Add space after `:`

File: business_requirement/models/business.py [hunk 12]
```diff
@@ -223,23 +225,77 @@ def project_id_change(self):
...
+                             'confirmation_date': False,
+                             'approval_date': False
+                             })
+            if vals['state'] == 'approved':
+                if br_xml_id.id in grps:
+                    vals.update({'approved_id': user,
+                                 'approval_date': date})
+                else:
+                    raise ValidationError(_(
+                        'You can only move to the following stage:'
```
Comment (seb-elico) [inline business_requirement/models/business.py]: If I'm correct, there wont be any space between `:` and `draft` in the message. Please add a space (same for other error messages)

**Review by seb-elico — CHANGES_REQUESTED** [review]

**Comment by YogeshMahera-SerpentCS** [comment]
@elicoidal 
I checked on runbot It's available

File: business_requirement/models/business.py (hunk 1 above)
Comment (YogeshMahera-SerpentCS) [reply business_requirement/models/business.py:199]: @victormartinelicocorp 
What are you refering to kanban_states or Kanban Stages?!

Comment (YogeshMahera-SerpentCS) [reply business_requirement/models/business.py:199]: Kanban states are visible in BR, I just sent the screenshot please check below.
@elicoidal @victormartinelicocorp

File: business_requirement/tests/test_br.py [hunk 13]
```diff
@@ -101,9 +104,29 @@ def test_br_name_search(self):
...
+        br_vals = {
+            'name': 'test',
+            'description': 'test',
+            'project_id': self.pr_1.id
+        }
+        self.br.create(br_vals)
+        br_vals1 = {
+            'project_id': self.pr_2.id
+        }
+        self.br.write(br_vals1)
```
Comment (martintorresvm) [inline business_requirement/tests/test_br.py]: None assert.

**Review by martintorresvm — CHANGES_REQUESTED** [review]
@YogeshMahera-SerpentCS @elicoidal @seb-elico 

All test was design just to pass coveralls. Should we do better now?

File: business_requirement/models/business.py (hunk 1 above)
Comment (martintorresvm) [reply business_requirement/models/business.py:199]: @YogeshMahera-SerpentCS my bad, please forgot about my comment.

**Comment by YogeshMahera-SerpentCS** [comment]
@victormartinelicocorp @seb-elico 
Could you please review

File: business_requirement/tests/test_br.py [hunk 14]
```diff
@@ -113,19 +113,21 @@ def test_create_sequance(self):
             'name': '/',
             'description': 'test',
         }
-        self.br.create(br_vals)
+        res = self.br.create(br_vals)
+        self.assertTrue(res.name)
```
Comment (martintorresvm) [inline business_requirement/tests/test_br.py]: @YogeshMahera-SerpentCS 
I want you to compare a sentence lets say:
`br_name = "BR0001/test"`
Then 
`self.assertEqual(br_name, res.name)`

File: business_requirement/tests/test_br.py [hunk 15]
```diff
@@ -113,19 +113,21 @@ def test_create_sequance(self):
...
             'project_id': self.pr_1.id
         }
-        self.br.create(br_vals)
+        res = self.br.create(br_vals)
         br_vals1 = {
             'project_id': self.pr_2.id
         }
-        self.br.write(br_vals1)
+        res = res.write(br_vals1)
+        self.assertTrue(res)
```
Comment (martintorresvm) [inline business_requirement/tests/test_br.py]: @YogeshMahera-SerpentCS Same like my previous comment

**Review by martintorresvm — CHANGES_REQUESTED** [review]

File: business_requirement/tests/test_br.py (hunk 14 above)
Comment (YogeshMahera-SerpentCS) [reply business_requirement/tests/test_br.py]: @victormartinelicocorp 
Sequance is automated created in create method i had tested but it's not Equal.
`vals['name'] = self.env['ir.sequence'].next_by_code(
                'business.requirement')`

Comment (YogeshMahera-SerpentCS) [reply business_requirement/tests/test_br.py]: @victormartinelicocorp I wish to compare it but I don't know what will be the next sequence number when test case will run. I cannot hard code the condition as you gave in the example.
Let me know if there is any way to compare it.

Comment (martintorresvm) [reply business_requirement/tests/test_br.py]: @YogeshMahera-SerpentCS 

Ok let's try from the beggining.

1. Test need to be called with the name of function you want cover.
2. Them from my point of view the name is wrong should be `test_create`
3. Analyse the function. 
4. Found one if condition then means you need cover that path and the case that is False. 
```
@api.model
def create(self, vals):
# From here is test_create_name_equal_slash
    if vals.get('name', '/') == '/':
        vals['name'] = self.env['ir.sequence'].get('business.requirement')
        vals['name'] = self.env['ir.sequence'].next_by_code(
            'business.requirement')
############################################

# From here is test_create_project_id
    if vals.get('project_id'):
        project_id = self.env['project.project'].\
        project_id = self.env['project.project']. \
            browse(vals.get('project_id'))
############################################

# From here is test_create_project_id_message_follower_ids

        if project_id and project_id.message_follower_ids:
            vals['message_follower_ids'] =\
                project_id.message_follower_ids.ids
            msg_followers = []
            for project in project_id.message_follower_ids:
                msg_vals = {
                    'channel_id': project.channel_id.id,
                    'display_name': project.display_name,
                    'partner_id': project.partner_id.id,
                    'res_model': self._name,
                    'subtype_ids': project.subtype_ids.ids
                }
                msg_followers.append((0, 0, msg_vals))
############################################

# From here is test_create_project_id_message_follower_ids_msg_followers
            if msg_followers:
                vals['message_follower_ids'] = msg_followers
############################################

    return super(BusinessRequirement, self).create(vals)
```

So the first test for me would be.
```
def test_create_name_equal_slash(self, vals):
    name = self.env['ir.sequence'].get('business.requirement')
    name = self.env['ir.sequence'].next_by_code(
        'business.requirement')
    br_vals = {
        'name': '/',
        'description': 'test',
    }
    res = self.br.create(br_vals)
    self.assertEqual(name, res.name)
```

This is my idea about how should be but open to suggestions

cc @elicoidal @seb-elico @sudhir-serpentcs

**Comment by YogeshMahera-SerpentCS** [comment]
@victormartinelicocorp Could you please review

File: business_requirement/report/report.xml [hunk 16]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <report
+            id="business_requirement_report"
```
Comment (seb-elico) [inline business_requirement/report/report.xml]: First parameter should be on the same line as the tag

File: business_requirement/views/business_view.xml [hunk 17]
```diff
@@ -1,329 +1,270 @@
...
+                           statusbar_visible="draft,confirmed,approved,stakeholder_approval,in_progress,done,cancel,drop"/>
+                </header>
+                <sheet>
+                    <div class="oe_left">
+                        <field name="kanban_state" class="oe_inline"
+                               widget="kanban_state_selection"/>
+                    </div>
+                    <div class="oe_right oe_button_box"
+                         name="buttons" groups="base.group_user">
+                         <button class="oe_inline oe_stat_button"
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Over-indented of 1 space

File: business_requirement/views/business_view.xml [hunk 18]
```diff
@@ -1,329 +1,270 @@
...
+                        <field name="origin"/>
+                        <field name="category_ids" widget="many2many_tags"/>
+                        <field name="to_be_reviewed"/>
+                        <field name="reviewer_ids" widget="many2many_tags"/>
+                        <field name="priority" groups="base.group_user" widget="priority"/>
+                    </group>
+                    <notebook>
+                        <page string="Business Requirement">
+                            <group string="Customer Story">
+                                 <field name="business_requirement" nolabel="1"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: All `field` tags below are over-indented of 1 space

File: business_requirement/views/business_view.xml [hunk 19]
```diff
@@ -1,329 +1,270 @@
...
+    <record id="view_business_requirement_category_form" model="ir.ui.view">
+        <field name="name">business.requirement.category.form</field>
+        <field name="model">business.requirement.category</field>
+        <field name="type">form</field>
+        <field name="priority">20</field>
+        <field name="arch" type="xml">
+            <form string="Business Requirement">
+                <sheet>
+                    <group>
+                         <field name="name"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Over-indented

File: business_requirement/views/business_view.xml [hunk 20]
```diff
@@ -1,329 +1,270 @@
...
-            </field>
-        </record>
+    <record id="view_business_requirement_category_tree" model="ir.ui.view">
+        <field name="name">business.requirement.category.form</field>
+        <field name="model">business.requirement.category</field>
+        <field name="type">tree</field>
+        <field name="priority">20</field>
+        <field name="arch" type="xml">
+            <tree string="Business Requirement">
+                 <field name="name"/>
```
Comment (seb-elico) [inline business_requirement/views/business_view.xml]: Over-indented

**Review by seb-elico — CHANGES_REQUESTED** [review]
Small indentation issues

**Comment by YogeshMahera-SerpentCS** [comment]
@seb-elico Improved Changes

**Review by seb-elico — APPROVED** [review]
LGTM

**Comment by martintorresvm** [comment]
@YogeshMahera-SerpentCS as is not clear then here is the work expected on UT:

Refactor UT from: business_requirement/tests/test_project.py:
- [ ] Fix name of file as is not true because all test are based on br object nothing is tested on project.
- [ ] Extract all test inside the `setUp` function (setUp is just a common part for each test).

Refactor UT from: business_requirement/tests/test_br.py 
- [ ] Check every function on b-r object and be sure to create the test for every condition.
      Follow example done on:
      - def test_create_name_equal_slash(self)
      - def test_create_write_project(self):

**Comment by elicoidal** [comment]
@YogeshMahera-SerpentCS let's try to finish it today :smile:

**Comment by YogeshMahera-SerpentCS** [comment]
@elicoidal Yes, I am doing my best.
Almost completed.

**Comment by YogeshMahera-SerpentCS** [comment]
@victormartinelicocorp Can you please review

**Review by martintorresvm — APPROVED** [review]
Let's move forward. The rest of UT to be improved on the next batch.

**Comment by bistaray** [comment]
After installing this as the first module on a new 10.0 Enterprise Edition database (no errors during installation), I can create a Category, but when I try to create a Business Requirement, I get:

TypeError: Cannot read property '0' of undefined

http://192.168.56.96:8069/web/content/262-96bcfcf/web.assets_backend.js:2258
Traceback:
TypeError: Cannot read property '0' of undefined
    at Class.prepare_dropdown_selection (http://192.168.56.96:8069/web/content/262-96bcfcf/web.assets_backend.js:2258:459)
    at Class.render_value (http://192.168.56.96:8069/web/content/262-96bcfcf/web.assets_backend.js:2259:92)
    at Class.prototype.(anonymous function) [as render_value] (http://192.168.56.96:8069/web/content/251-9a56ee6/web.assets_common.js:3009:559)
    at Class.start (http://192.168.56.96:8069/web/content/262-96bcfcf/web.assets_backend.js:2235:106)
    at Class.prototype.(anonymous function) [as start] (http://192.168.56.96:8069/web/content/251-9a56ee6/web.assets_common.js:3009:559)
    at Class.start (http://192.168.56.96:8069/web/content/262-96bcfcf/web.assets_backend.js:2258:333)
    at Class.prototype.(anonymous function) [as start] (http://192.168.56.96:8069/web/content/251-9a56ee6/web.assets_common.js:3009:559)
    at http://192.168.56.96:8069/web/content/251-9a56ee6/web.assets_common.js:3112:963
    at http://192.168.56.96:8069/web/content/251-9a56ee6/web.assets_common.js:547:681
    at fire (http://192.168.56.96:8069/web/content/251-9a56ee6/web.assets_common.js:541:299)

And am blocked from proceeding, even after this message is dismissed.

**Comment by elicoidal** [comment]
@bistaray thanks for reporting: I could not reproduce the error.
I created a new category and then tried to create a BR and did not experience any issue.
Do you have a pattern to reproduce it? Could you input an issue if you can reproduce it with the steps?
Thanks!

**Comment by vcloud9ray** [comment]
@elicoidal - I pulled the latest version of Odoo 10 Community + Enterprise and re-downloaded the repo. Now I can create a BR without the error. So whatever my issue was, it is gone now. (bistaray -> vcloud9raymond)

**Comment by elicoidal** [comment]
@vcloud9raymond Great! thanks for reporting back

**Pull request merged**